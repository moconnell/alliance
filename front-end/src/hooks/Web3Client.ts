import Web3Modal from "web3modal";
import WalletConnectProvider from "@walletconnect/web3-provider";
import { useEffect, useReducer, useCallback, useMemo } from "react";
import { ethers } from "ethers";
import { Web3ProviderState, web3InitialState, web3Reducer } from "../reducers";

const getWeb3Modal = () => {
  if (typeof window === "undefined") return null;

  const {
    REACT_APP_WEB3_CACHE_PROVIDER,
    REACT_APP_WEB3_INFURA_ID,
    REACT_APP_WEB3_NETWORK,
  } = process.env;

  return new Web3Modal({
    network: REACT_APP_WEB3_NETWORK,
    cacheProvider: REACT_APP_WEB3_CACHE_PROVIDER,
    providerOptions: {
      walletconnect: {
        package: WalletConnectProvider,
        options: {
          infuraId: REACT_APP_WEB3_INFURA_ID,
        },
      },
    },
  });
};

export type Web3Client = Web3ProviderState & {
  connect: () => Promise<void>;
  disconnect: () => Promise<void>;
};

export const useWeb3 = () => {
  const [state, dispatch] = useReducer(web3Reducer, web3InitialState);
  const { provider, web3Provider, address, network, signer } = state;
  const web3Modal = useMemo(() => getWeb3Modal(), []);

  const connect = useCallback(async () => {
    try {
      if (!web3Modal) {
        console.error("Cannot connect: no Web3Modal");
        return;
      }

      const provider = await web3Modal.connect();
      const web3Provider = new ethers.providers.Web3Provider(provider);
      const signer = web3Provider.getSigner();
      const address = await signer.getAddress();
      const network = await web3Provider.getNetwork();

      dispatch({
        type: "SET_WEB3_PROVIDER",
        provider,
        web3Provider,
        address,
        network,
        signer,
      });
    } catch (e) {
      console.log("connect error", e);
    }
  }, [web3Modal]);

  const disconnect = useCallback(async () => {
    if (web3Modal) {
      web3Modal.clearCachedProvider();
      if (provider?.disconnect && typeof provider.disconnect === "function") {
        await provider.disconnect();
      }
      dispatch({
        type: "RESET_WEB3_PROVIDER",
      });
    } else {
      console.error("No Web3Modal");
    }
  }, [provider, web3Modal]);

  // Auto connect to the cached provider
  useEffect(() => {
    if (web3Modal && web3Modal.cachedProvider) {
      connect();
    }
  }, [connect, web3Modal]);

  // EIP-1193 events
  useEffect(() => {
    if (provider?.on) {
      const handleAccountsChanged = (accounts: string[]) => {
        dispatch({
          type: "SET_ADDRESS",
          address: accounts[0],
        });
      };

      // https://docs.ethers.io/v5/concepts/best-practices/#best-practices--network-changes
      const handleChainChanged = (_hexChainId: string) => {
        if (typeof window !== "undefined") {
          console.log("switched to chain...", _hexChainId);
          window.location.reload();
        } else {
          console.log("window is undefined");
        }
      };

      const handleDisconnect = (error: { code: number; message: string }) => {
        // eslint-disable-next-line no-console
        console.log("disconnect", error);
        disconnect();
      };

      provider.on("accountsChanged", handleAccountsChanged);
      provider.on("chainChanged", handleChainChanged);
      provider.on("disconnect", handleDisconnect);

      // Subscription Cleanup
      return () => {
        if (provider.removeListener) {
          provider.removeListener("accountsChanged", handleAccountsChanged);
          provider.removeListener("chainChanged", handleChainChanged);
          provider.removeListener("disconnect", handleDisconnect);
        }
      };
    }
  }, [provider, disconnect]);

  return {
    provider,
    web3Provider,
    address,
    network,
    signer,
    connect,
    disconnect,
  } as Web3Client;
};
