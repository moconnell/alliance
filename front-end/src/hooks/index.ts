export type { Web3Client } from "./Web3Client";
export { useWeb3 } from "./Web3Client";
export { useCalendar } from "./useCalendar";
