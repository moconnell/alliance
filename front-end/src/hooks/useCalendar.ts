import { useCallback, useEffect, useState } from "react";

import { useWeb3Context } from "../context/Web3Context";
import assert from "assert";
import {
  Calendar,
  CalendarFactory,
  CalendarFactory__factory,
  Calendar__factory,
} from "../../typechain-types";
import DaysOfWeek from "../types/daysOfWeek";
import {
  assertValid,
  compare,
  fromTotalMins,
  totalMinutes,
} from "../utils/timeUtils";
import Meeting from "../types/meeting";
import ProfileInfo from "../types/profileInfo";
import AvailabilityInfo from "../types/availabilityInfo";
import { DateTime } from "luxon";

const NULL_ADDRESS = "0x0000000000000000000000000000000000000000";

const toString = (value: any) => (value ? String(value) : "");

export const useCalendar = () => {
  const { address, signer, web3Provider } = useWeb3Context();
  const [calendarFactory, setCalendarFactory] = useState<CalendarFactory>();
  const [calendar, setCalendar] = useState<Calendar>();
  const [availability, setAvailability] = useState<AvailabilityInfo>();
  const [profile, setProfile] = useState<ProfileInfo>();

  const { REACT_APP_WEB3_CONTRACT_FACTORY_ADDRESS } = process.env;
  assert(
    REACT_APP_WEB3_CONTRACT_FACTORY_ADDRESS,
    "REACT_APP_WEB3_CONTRACT_FACTORY_ADDRESS not set"
  );

  useEffect(() => {
    function getCalendarFactory() {
      if (!web3Provider || !address || !signer) return undefined;
      return CalendarFactory__factory.connect(
        REACT_APP_WEB3_CONTRACT_FACTORY_ADDRESS!,
        signer
      );
    }

    const fac = getCalendarFactory();
    if (fac) setCalendarFactory(fac);
  }, [REACT_APP_WEB3_CONTRACT_FACTORY_ADDRESS, address, signer, web3Provider]);

  useEffect(() => {
    async function getCalendar() {
      if (!address || !calendarFactory || !web3Provider) return;
      console.log(calendarFactory);
      const calendarAddress = await calendarFactory.userToCalendar(address);
      if (calendarAddress === NULL_ADDRESS) {
        console.log(`wallet ${address} has no saved calendar`);
        return;
      }

      console.log(`fetching calendar at ${calendarAddress}`);

      const cal = Calendar__factory.connect(calendarAddress, web3Provider);
      setCalendar(cal);
    }

    getCalendar();
  }, [address, calendarFactory, web3Provider]);

  useEffect(() => {
    async function loadProfileAvailability() {
      if (!calendar) return;

      const newProfile = await calendar.profile();
      setProfile(newProfile);

      let newAvailability = await calendar.availability();
      const from = fromTotalMins(newAvailability.earliestStartMinutes);
      const to = fromTotalMins(
        newAvailability.earliestStartMinutes + newAvailability.minutesAvailable
      );
      setAvailability({ ...newAvailability, from, to });
    }

    loadProfileAvailability();
  }, [calendar]);

  useEffect(() => {
    const createCalendar = async () => {
      if (
        calendar ||
        !calendarFactory ||
        !web3Provider ||
        !profile ||
        !availability
      )
        return;

      assert(profile.email, "No email");
      assertValid(availability.from);
      assertValid(availability.to);
      assert(
        compare(availability.from!, availability.to!) < 0,
        "from after to"
      );

      const p = {
        description: toString(profile.description),
        email: profile.email!,
        picture: toString(profile.picture),
        url: toString(profile.url),
        username: toString(profile.username),
      };

      const a = {
        availableDays: availability.availableDays ?? DaysOfWeek.None,
        earliestStartMinutes: totalMinutes(availability.from!),
        minutesAvailable:
          totalMinutes(availability.to!) - totalMinutes(availability.from!),
        location: availability.location!,
        timeZone: availability.timeZone!,
      };

      console.log("creating calendar now ... ", p, a);

      const tx = await calendarFactory.createCalendar(p, a);
      const receipt = await tx.wait();
      const txEvent = receipt.events?.[0]?.args;
      const calendarAddr = txEvent?.calendar;
      const cal = Calendar__factory.connect(calendarAddr, web3Provider);

      console.log("calendar created ok", cal);

      setCalendar(cal);
    };

    createCalendar();
  }, [
    profile,
    availability,
    calendar,
    calendarFactory,
    web3Provider,
    address,
    REACT_APP_WEB3_CONTRACT_FACTORY_ADDRESS,
  ]);

  const getMeetings = useCallback(
    async (date: Date) => {
      if (!calendar) return undefined;

      console.log(`getMeetings(date: ${date})`);

      const year = date.getFullYear();
      const month = date.getMonth() + 1;
      const day = date.getDate();
      const meetingsOnDate = await calendar.getMeetings(year, month, day);

      return meetingsOnDate.map(
        ({ attendee, startMinutes, durationMinutes }) => {
          const hour = Math.floor(startMinutes / 60);
          const minute = startMinutes % 60;
          const from = DateTime.fromObject(
            {
              year,
              month,
              day,
              hour,
              minute,
            },
            {
              zone: availability?.timeZone,
            }
          ).setZone();
          const to = from.plus({ minutes: durationMinutes });

          return {
            date: from,
            endDate: to,
            attendee: attendee,
            description: `${durationMinutes} minute meeting`,
          } as Meeting;
        }
      );
    },
    [availability?.timeZone, calendar]
  );

  const getProfileAvailability = useCallback(
    async (calendarAddress: string) => {
      if (!web3Provider) return undefined;

      console.log(
        `getProfileAvailability(calendarAddress: ${calendarAddress})`
      );

      const otherCalendar = Calendar__factory.connect(
        calendarAddress,
        web3Provider
      );

      const profileAvailability: [ProfileInfo, AvailabilityInfo] = [
        await otherCalendar.profile(),
        await otherCalendar.availability(),
      ];

      return profileAvailability;
    },
    [web3Provider]
  );

  const getAvailableTimes = useCallback(
    async (calendarAddress: string, date: Date, durationMinutes: number) => {
      if (!web3Provider) return undefined;

      console.log(
        `getAvailableTimes(calendarAddress: ${calendarAddress}, date: ${date}, duration: ${durationMinutes})`
      );

      const otherCalendar = Calendar__factory.connect(
        calendarAddress,
        web3Provider
      );

      const year = date.getFullYear();
      const month = date.getMonth() + 1;
      const day = date.getDate();
      const { times, timeZone } = await otherCalendar.getAvailableTimes(
        year,
        month,
        day,
        durationMinutes
      );

      return times
        .filter((startMinutes) => startMinutes > 0)
        .map((startMinutes) => {
          const hour = Math.floor(startMinutes / 60),
            minute = startMinutes % 60;

          return DateTime.fromObject(
            { year, month, day, hour, minute },
            { zone: timeZone }
          );
        });
    },
    [web3Provider]
  );

  const bookMeeting = useCallback(
    async (calendarAddress: string, at: DateTime, durationMinutes: number) => {
      if (!signer) return;

      console.log(
        `bookMeeting(calendarAddress: ${calendarAddress}, at: ${at}, duration: ${durationMinutes})`
      );

      const otherCalendar = Calendar__factory.connect(calendarAddress, signer);

      await otherCalendar.bookMeeting(
        at.year,
        at.month,
        at.day,
        at.hour,
        at.minute,
        durationMinutes
      );
    },
    [signer]
  );

  const setProfileAvailability = (info: ProfileInfo & AvailabilityInfo) => {
    setProfile(info);
    setAvailability(info);
  };

  return {
    address,
    availability,
    bookMeeting,
    getAvailableTimes,
    getMeetings,
    getProfileAvailability,
    hasCalendar: calendar !== undefined,
    profile,
    setProfileAvailability,
  };
};
