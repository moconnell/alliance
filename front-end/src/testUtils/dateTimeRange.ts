import { DateTime } from "luxon";
import { range } from "../utils/range";

export const dateTimeRange = (
  start: number,
  end: number,
  intervalMinutes: 60 | 30 | 20 | 15 | 10 | 5 = 60,
  locale: string = "en-us"
) =>
  range(start, end).flatMap((hour) =>
    range(0, 60 / intervalMinutes - 1).map((i) =>
      DateTime.fromObject({ hour, minute: i * intervalMinutes }).setLocale(
        locale
      )
    )
  );
