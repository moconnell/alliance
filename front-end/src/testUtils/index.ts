export { dateTimeRange } from "./dateTimeRange";
export { renderWithFormProvider } from "./reactHookFormUtils";
