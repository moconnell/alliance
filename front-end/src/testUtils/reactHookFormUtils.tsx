import { render } from "@testing-library/react";
import { useForm, FormProvider } from "react-hook-form";

interface WithFormProviderProps {
  children: React.ReactNode;
  defaultValues?: {
    [x: string]: any;
  };
}

const WithFormProvider = ({
  children,
  defaultValues,
}: WithFormProviderProps) => {
  const methods = useForm(defaultValues);

  return <FormProvider {...methods}>{children}</FormProvider>;
};

export const renderWithFormProvider = <
  T extends {
    [x: string]: any;
  }
>(
  control: JSX.Element,
  defaultValues?: T
) => {
  return render(
    <WithFormProvider defaultValues={defaultValues}>{control}</WithFormProvider>
  );
};
