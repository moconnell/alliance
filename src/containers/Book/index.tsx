import {
  Button,
  Container,
  Flex,
  Heading,
  Table,
  Text,
  Tbody,
  Td,
  Th,
  Tr,
  useToast,
} from "@chakra-ui/react";
import { DateTime } from "luxon";
import { Fragment, useEffect, useState } from "react";
import Calendar from "../../components/Calendar";
import { useParams } from "react-router-dom";
import { useCalendar } from "../../hooks";
import DaysOfWeek from "../../types/daysOfWeek";
import { tomorrow } from "../../utils/timeUtils";
import ModalComponent from "../../components/Modal";
import LoadingTransaction from "../../components/Modal/components/LoadingTransaction";
import TimeList from "../../components/TimeList";
import { isAvailable } from "../../utils/daysOfWeekUtils";

enum BookingState {
  None,
  Requested,
  InProgress,
  Complete,
}

const Book = () => {
  const { calendarAddress } = useParams();
  const { address, bookMeeting, getProfileAvailability, getAvailableTimes } =
    useCalendar();
  const [selectedDate, setSelectedDate] = useState(tomorrow());
  const [selectedTime, setSelectedTime] = useState<DateTime | undefined>();
  const [durationMinutes] = useState(60);
  const [availableDays, setAvailableDays] = useState(DaysOfWeek.None);
  const [availableTimes, setAvailableTimes] = useState([] as DateTime[]);
  const [description, setDescription] = useState("");
  const [location, setLocation] = useState("");
  const [timeZone, setTimeZone] = useState("");
  const [username, setUsername] = useState("");
  const [showModal, setShowModal] = useState(false);
  const [bookingState, setBookingState] = useState(BookingState.None);
  const toast = useToast();
  const now = new Date();

  useEffect(() => {
    const loadProfileAvailability = async () => {
      if (!calendarAddress) return;
      const profileAvailability = await getProfileAvailability(calendarAddress);
      if (profileAvailability) {
        const [
          { username, description },
          { availableDays, location, timeZone },
        ] = profileAvailability;
        if (availableDays) setAvailableDays(availableDays);
        if (location) setLocation(location);
        if (timeZone) setTimeZone(timeZone);
        if (username) setUsername(username);
        if (description) setDescription(description);
      }
    };

    loadProfileAvailability();
  }, [calendarAddress, getProfileAvailability]);

  useEffect(() => {
    const loadTimes = async () => {
      if (!calendarAddress) return;
      const availableTimes_ = await getAvailableTimes(
        calendarAddress,
        selectedDate,
        durationMinutes
      );
      if (availableTimes_) {
        setAvailableTimes(availableTimes_);
      }
    };

    loadTimes();
  }, [calendarAddress, selectedDate, durationMinutes, getAvailableTimes]);

  useEffect(() => {
    const doBooking = async () => {
      if (
        bookingState === BookingState.Requested &&
        calendarAddress &&
        selectedTime &&
        durationMinutes
      ) {
        try {
          await bookMeeting(calendarAddress, selectedTime, durationMinutes);
          setBookingState(BookingState.Complete);
        } catch (error) {
          toast({
            id: "booking-error",
            status: "error",
            title: "Booking Error",
            description: String(error),
            duration: 10,
            isClosable: true,
          });
          setBookingState(BookingState.None);
        }
      }
    };

    doBooking();
  }, [
    bookMeeting,
    bookingState,
    calendarAddress,
    selectedTime,
    durationMinutes,
    toast,
  ]);

  function handleCloseModal() {
    setShowModal(false);
  }

  if (!calendarAddress?.length || !durationMinutes)
    return (
      <Container data-testid="container:invalid-url" maxW="container.xl" p={1}>
        <Heading as="h1" color="raid.100" fontFamily="Mirza,serif">
          Hmm...
        </Heading>
        <Text fontFamily="Inter" color="#fff">
          That looks like an invalid link - please check and try again!
        </Text>
      </Container>
    );

  if (address === calendarAddress)
    return (
      <Container
        data-testid="container:invalid-calendar"
        maxW="container.xl"
        p={1}
      >
        <Heading as="h1" color="raid.100" fontFamily="Mirza,serif">
          Hey, that's you!
        </Heading>
        <Text fontFamily="Inter" color="#fff">
          Did you really want to book a meeting with yourself?
        </Text>
      </Container>
    );

  switch (bookingState) {
    case BookingState.Complete:
      return (
        <Container
          data-testid="container:booking-complete"
          maxW="container.xl"
          p={1}
        >
          <Heading as="h1" color="raid.100" fontFamily="Mirza,serif">
            Success!
          </Heading>
          <Text fontFamily="Inter" color="#fff">
            {`Your ${durationMinutes} minute meeting with ${username} is all
              booked for ${selectedTime
                ?.toLocal()
                .toLocaleString({ dateStyle: "full", timeStyle: "short" })}.`}
          </Text>
        </Container>
      );
    default:
      return (
        <Container data-testid="container:book" maxW="container.xl" p={1}>
          <Heading as="h1" color="raid.100" fontFamily="Mirza,serif">
            Book a {durationMinutes} minute meeting
            {username && ` with ${username}`}
            ...
          </Heading>
          <Flex
            h={{ base: "auto", md: "100%" }}
            py={10}
            direction={{ base: "column", md: "row" }}
          >
            {(location || timeZone || description) && (
              <Container data-testid="container:profile" paddingTop="24">
                <Table>
                  <Tbody fontFamily="Inter">
                    {location && (
                      <Tr>
                        <Th border={0} color="gray.500">
                          Location
                        </Th>
                        <Td border={0} color="gray.300" fontSize="sm">
                          {location}
                        </Td>
                      </Tr>
                    )}
                    {timeZone && (
                      <Tr>
                        <Th border={0} color="gray.500">
                          Time-Zone
                        </Th>
                        <Td border={0} color="gray.300" fontSize="sm">
                          {timeZone}
                        </Td>
                      </Tr>
                    )}
                    {description && (
                      <Tr>
                        <Th border={0} color="gray.500">
                          About
                        </Th>
                        <Td border={0} color="gray.300" fontSize="sm">
                          {description}
                        </Td>
                      </Tr>
                    )}
                  </Tbody>
                </Table>
              </Container>
            )}
            <Container p={0}>
              <Calendar
                onChange={setSelectedDate}
                defaultValue={selectedDate}
                tileDisabled={(date) =>
                  date < now || !isAvailable(date, availableDays)
                }
              />
            </Container>
            <Container p={4}>
              <TimeList
                selectedTime={selectedTime}
                onChange={setSelectedTime}
                times={availableTimes}
                duration={durationMinutes}
              />
            </Container>
          </Flex>
          <Container pr={10} maxW="xs">
            <Button
              data-testid="button:book-meeting"
              disabled={
                !selectedDate ||
                !selectedTime ||
                bookingState !== BookingState.None
              }
              onClick={() => setBookingState(BookingState.Requested)}
            >
              Book meeting
            </Button>
          </Container>
          <ModalComponent showModal={showModal} closeModal={handleCloseModal}>
            <LoadingTransaction />
          </ModalComponent>
        </Container>
      );
  }
};

export default Book;
