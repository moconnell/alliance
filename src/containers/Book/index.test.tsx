import { fireEvent, render, waitFor } from "@testing-library/react";
import { createMemoryHistory } from "history";
import { DateTime } from "luxon";
import { Route, Router, Routes } from "react-router-dom";
import DaysOfWeek from "../../types/daysOfWeek";
import { dateTimeRange } from "../../testUtils";
import { useCalendar } from "../../hooks";
import AvailabilityInfo from "../../types/availabilityInfo";
import ProfileInfo from "../../types/profileInfo";
import Book from "./index";
import { act } from "react-dom/test-utils";

const HARDHAT_ACC_0 = "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266";
const HARDHAT_ACC_1 = "0x23dB4a08f2272df049a4932a4Cc3A6Dc1002B33E";
const MR_BOJANGLES_PROFILE_AVAILABILITY: [ProfileInfo, AvailabilityInfo] = [
  { username: "mr_bojangles", description: "legend" },
  {
    location: "The Big Smoke",
    timeZone: "Europe/London",
    availableDays: DaysOfWeek.MonFri,
  },
];

jest.mock("../../hooks");
jest.mock("../../components/Calendar", () => () => "((Calendar))");
jest.mock("../../components/Modal", () => () => "((ModalComponent))");
jest.mock(
  "../../components/Modal/components/LoadingTransaction",
  () => () => "((LoadingTransaction))"
);

describe("<Book />", () => {
  const mockBookMeeting = jest.fn();
  const mockGetProfileAvailability = jest.fn();
  const mockGetAvailableTimes = jest.fn();

  const renderWithRouter = (calendarAddress?: string) => {
    const history = createMemoryHistory();
    history.push(calendarAddress ? `/book/${calendarAddress}` : "/book");

    return render(
      <Router location={history.location} navigator={history}>
        <Routes>
          <Route path="/book" element={<Book />}>
            <Route path="/book/:calendarAddress" element={<Book />} />
          </Route>
        </Routes>
      </Router>
    );
  };

  afterEach(() => jest.resetAllMocks());
  afterAll(() => jest.restoreAllMocks());

  const testData: [
    string,
    string,
    string?,
    string?,
    [ProfileInfo, AvailabilityInfo]?,
    DateTime[]?
  ][] = [
    [
      "render correctly",
      "container:profile",
      HARDHAT_ACC_0,
      HARDHAT_ACC_1,
      MR_BOJANGLES_PROFILE_AVAILABILITY,
      dateTimeRange(9, 14),
    ],
    ["render invalid url error", "container:invalid-url"],
    [
      "render invalid calendar error",
      "container:invalid-calendar",
      HARDHAT_ACC_0,
      HARDHAT_ACC_0,
    ],
  ];

  testData.forEach(
    ([
      name,
      testId,
      address,
      calendarAddress,
      profileAvailability,
      availableTimes,
    ]) =>
      it(`should ${name}`, async () => {
        (useCalendar as jest.Mock).mockImplementation(() => {
          return {
            address,
            bookMeeting: mockBookMeeting,
            getProfileAvailability: mockGetProfileAvailability,
            getAvailableTimes: mockGetAvailableTimes,
          };
        });

        mockGetProfileAvailability.mockResolvedValue(profileAvailability);
        mockGetAvailableTimes.mockResolvedValue(availableTimes);

        const { asFragment, getByTestId } = renderWithRouter(calendarAddress);

        await waitFor(() => expect(getByTestId(testId)).toBeInTheDocument());

        expect(asFragment()).toMatchSnapshot();
      })
  );

  it("should book meeting", async () => {
    (useCalendar as jest.Mock).mockImplementation(() => {
      return {
        address: HARDHAT_ACC_0,
        bookMeeting: mockBookMeeting,
        getProfileAvailability: mockGetProfileAvailability,
        getAvailableTimes: mockGetAvailableTimes,
      };
    });

    mockGetProfileAvailability.mockResolvedValue(
      MR_BOJANGLES_PROFILE_AVAILABILITY
    );
    const mockTimes = dateTimeRange(9, 17, 30);
    mockGetAvailableTimes.mockResolvedValue(mockTimes);

    const { getByTestId, getByText } = renderWithRouter(HARDHAT_ACC_1);

    await waitFor(() =>
      expect(getByTestId("container:profile")).toBeInTheDocument()
    );

    act(() => {
      fireEvent.click(getByText("10:30 AM"));
    });

    await waitFor(() =>
      expect(getByTestId("button:book-meeting")).toBeEnabled()
    );

    act(() => {
      fireEvent.click(getByTestId("button:book-meeting"));
    });

    await waitFor(() =>
      expect(getByTestId("container:booking-complete")).toBeInTheDocument()
    );

    expect(mockBookMeeting).toHaveBeenCalledWith(
      HARDHAT_ACC_1,
      mockTimes[3],
      60
    );
  });

  it("should show toast on bookMeeting error", async () => {
    (useCalendar as jest.Mock).mockImplementation(() => {
      return {
        address: HARDHAT_ACC_0,
        bookMeeting: mockBookMeeting,
        getProfileAvailability: mockGetProfileAvailability,
        getAvailableTimes: mockGetAvailableTimes,
      };
    });

    mockGetProfileAvailability.mockResolvedValue(
      MR_BOJANGLES_PROFILE_AVAILABILITY
    );

    const mockTimes = dateTimeRange(9, 17, 30);
    mockGetAvailableTimes.mockResolvedValue(mockTimes);

    const errorMessage = "Ooops, looks like something went wrong there :/";
    mockBookMeeting.mockImplementation(() => {
      throw new Error(errorMessage);
    });

    const { getByRole, getByTestId, getByText } =
      renderWithRouter(HARDHAT_ACC_1);

    await waitFor(() =>
      expect(getByTestId("container:profile")).toBeInTheDocument()
    );

    act(() => {
      fireEvent.click(getByText("10:30 AM"));
    });

    await waitFor(() =>
      expect(getByTestId("button:book-meeting")).toBeEnabled()
    );

    act(() => {
      fireEvent.click(getByTestId("button:book-meeting"));
    });

    await waitFor(() => expect(getByRole("alert")).toBeInTheDocument());

    expect(getByRole("alert").innerHTML).toContain(errorMessage);
  });
});
