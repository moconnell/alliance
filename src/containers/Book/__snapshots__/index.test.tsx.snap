// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Book /> should render correctly 1`] = `
<DocumentFragment>
  <div
    class="chakra-container css-122pq3t"
    data-testid="container:book"
  >
    <h1
      class="chakra-heading css-1lg6lob"
    >
      Book a 60 minute meeting with mr_bojangles...
    </h1>
    <div
      class="css-b82b48"
    >
      <div
        class="chakra-container css-19n8dai"
        data-testid="container:profile"
      >
        <table
          class="chakra-table css-0"
          role="table"
        >
          <tbody
            class="css-13c8hy8"
          >
            <tr
              class="css-0"
              role="row"
            >
              <th
                class="css-1u3nbe2"
              >
                Location
              </th>
              <td
                class="css-17arw3k"
                role="gridcell"
              >
                The Big Smoke
              </td>
            </tr>
            <tr
              class="css-0"
              role="row"
            >
              <th
                class="css-1u3nbe2"
              >
                Time-Zone
              </th>
              <td
                class="css-17arw3k"
                role="gridcell"
              >
                Europe/London
              </td>
            </tr>
            <tr
              class="css-0"
              role="row"
            >
              <th
                class="css-1u3nbe2"
              >
                About
              </th>
              <td
                class="css-17arw3k"
                role="gridcell"
              >
                legend
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <div
        class="chakra-container css-old1by"
      >
        ((Calendar))
      </div>
      <div
        class="chakra-container css-2rlxtj"
      >
        <div
          class="sc-eCYdqJ ilNZsE"
        >
          <div
            class="sc-jSMfEi Teyip"
          >
            Select a time
          </div>
          <div
            class="sc-gKXOVf itiFUh"
          >
            <div
              class="sc-gKXOVf itiFUh"
            >
              <svg>
                clock.svg
              </svg>
            </div>
            <span
              class="sc-dkzDqf jtoqIZ"
            >
              60 mins
            </span>
          </div>
          <div
            class="sc-gsnTZi efgEfA"
          >
            <button
              class="sc-bczRLJ fYCUyx"
              style="background-color: transparent;"
            >
              9:00 AM
            </button>
            <button
              class="sc-bczRLJ fYCUyx"
              style="background-color: transparent;"
            >
              10:00 AM
            </button>
            <button
              class="sc-bczRLJ fYCUyx"
              style="background-color: transparent;"
            >
              11:00 AM
            </button>
            <button
              class="sc-bczRLJ fYCUyx"
              style="background-color: transparent;"
            >
              12:00 PM
            </button>
            <button
              class="sc-bczRLJ fYCUyx"
              style="background-color: transparent;"
            >
              1:00 PM
            </button>
            <button
              class="sc-bczRLJ fYCUyx"
              style="background-color: transparent;"
            >
              2:00 PM
            </button>
          </div>
          <div
            class="sc-hKMtZM evkSWm"
          >
            Times shown in local time
          </div>
        </div>
      </div>
    </div>
    <div
      class="chakra-container css-5exlul"
    >
      <button
        class="chakra-button css-5c2176"
        data-testid="button:book-meeting"
        disabled=""
        type="button"
      >
        Book meeting
      </button>
    </div>
    ((ModalComponent))
  </div>
</DocumentFragment>
`;

exports[`<Book /> should render invalid calendar error 1`] = `
<DocumentFragment>
  <div
    class="chakra-container css-122pq3t"
    data-testid="container:invalid-calendar"
  >
    <h1
      class="chakra-heading css-1lg6lob"
    >
      Hey, that's you!
    </h1>
    <p
      class="chakra-text css-170nc93"
    >
      Did you really want to book a meeting with yourself?
    </p>
  </div>
</DocumentFragment>
`;

exports[`<Book /> should render invalid url error 1`] = `
<DocumentFragment>
  <div
    class="chakra-container css-122pq3t"
    data-testid="container:invalid-url"
  >
    <h1
      class="chakra-heading css-1lg6lob"
    >
      Hmm...
    </h1>
    <p
      class="chakra-text css-170nc93"
    >
      That looks like an invalid link - please check and try again!
    </p>
  </div>
</DocumentFragment>
`;
